Source/WebCore/dom/InternalObserverReduce.cpp

```diff
@@ -0,0 +1,145 @@
+#include "SubscribeOptions.h"
+#include "Subscriber.h"
+#include "SubscriberCallback.h"
+#include <JavaScriptCore/JSCJSValueInlines.h>
+
+namespace WebCore {
+
+class InternalObserverReduce final : public InternalObserver {
+public:
+    static Ref<InternalObserverReduce> create(ScriptExecutionContext& context, Ref<AbortSignal>&& signal, Ref<ReducerCallback>&& callback, std::optional<JSC::JSValue>&& initialValue, Ref<DeferredPromise>&& promise)
+    {
+        Ref internalObserver = adoptRef(*new InternalObserverReduce(context, WTFMove(signal), WTFMove(callback), WTFMove(initialValue), WTFMove(promise)));
+        internalObserver->suspendIfNeeded();
+        return internalObserver;
+    }
+
+private:
+    void next(JSC::JSValue value) final
+    {
+        if (!m_accumulator) {
+            m_idx++;
+            m_accumulator = value;
+            return;
+        }
+
+        auto* globalObject = protectedScriptExecutionContext()->globalObject();
+        ASSERT(globalObject);
+
+        Ref vm = globalObject->vm();
+
+        JSC::JSLockHolder lock(vm);
+        auto scope = DECLARE_CATCH_SCOPE(vm);
+
+        auto result = protectedCallback()->handleEventRethrowingException(*m_accumulator, value, m_idx++);
+
+        JSC::Exception* exception = scope.exception();
+        if (UNLIKELY(exception)) {
+            scope.clearException();
+            auto value = exception->value();
+            protectedPromise()->reject<IDLAny>(value);
+            Ref { m_signal }->signalAbort(value);
+        }
+
+        if (result.type() == CallbackResultType::Success)
+            m_accumulator = result.releaseReturnValue();
+    }
+
+    void error(JSC::JSValue value) final
+    {
+        protectedPromise()->reject<IDLAny>(value);
+    }
+
+    void complete() final
+    {
+        InternalObserver::complete();
+
+        if (!m_accumulator) {
+            protectedPromise()->reject(Exception { ExceptionCode::TypeError, "No inital value for Observable with no values"_s });
+            return;
+        }
+
+        protectedPromise()->resolve<IDLAny>(m_accumulator.value());
+    }
+
+    void visitAdditionalChildren(JSC::AbstractSlotVisitor& visitor) const final
+    {
+        protectedCallback()->visitJSFunction(visitor);
+    }
+
+    Ref<DeferredPromise> protectedPromise() const { return m_promise; }
+    Ref<ReducerCallback> protectedCallback() const { return m_callback; }
+
+    InternalObserverReduce(ScriptExecutionContext& context, Ref<AbortSignal>&& signal, Ref<ReducerCallback>&& callback, std::optional<JSC::JSValue>&& initialValue, Ref<DeferredPromise>&& promise)
+        : InternalObserver(context)
+        , m_signal(WTFMove(signal))
+        , m_callback(WTFMove(callback))
+        , m_accumulator(WTFMove(initialValue))
+        , m_promise(WTFMove(promise))
+    {
+    }
+
+    uint64_t m_idx { 0 };
+    Ref<AbortSignal> m_signal;
+    Ref<ReducerCallback> m_callback;
+    std::optional<JSC::JSValue> m_accumulator;
```

Comment:
@darinadler @cdumez Should this really be

```suggestion
    std::optional<JSValueInWrappedObject> m_accumulator;
```

over slack you mentioned that with like with [last](https://github.com/WebKit/WebKit/blob/a542f6c9c8b1f812fe86dea9055334e1ddf4a0d1/Source/WebCore/dom/InternalObserverLast.cpp#L86) nothing is keeping the value alive?

Reply:
Oh, yes this is wrong, can’t just put a JSValue in an object without wrapping it. It’s not enough to just use `JSValueInWrappedObject`.

Reply:
What would you suggest as per a fix?

We have the visitAdditionalChildren wirings already, and I see `JSValueInWrappedObject` has a `visit` method, can I use that and visit in our existing visitAdditionalChildren?

Reply:
Yes, I think so …

Reply:
There is a comment that kind of makes me want to use `set`, or rather when would it be "cleared from under you" as per the setWeakly (default) behaviour?

https://github.com/WebKit/WebKit/blob/c112fe9adf5d1c10ca6abf2542900dda02286f03/Source/WebCore/bindings/js/JSValueInWrappedObject.h#L49-L51

If I should be using `set`, where would I get a `JSC::JSCell`. Most of the sample code I can find around, simply use the `setWeakly` (default) behaviour, so maybe it is okay.

Reply:
Perhaps the JSCell (owner argument), can be the `Observable` instance object?

Reply:
Oh wait never mind, we are calling `visit`, in our visitAdditionalChildren — so it wont be cleaned from under us, and its safe.

Reply:
Raised PR fix for the `last` operator: https://github.com/WebKit/WebKit/pull/37670